src/context/stack.rs

```diff
@@ -0,0 +1,195 @@
+pub(super) struct Stack {
+    pub(super) stack_bottom: *mut c_void,
+    // high 48 bit means version, lower 16 bits are id
+    pub(super) stack_id: u64,
+    stack_size: usize,
+}
+
+impl Stack {
+    #[cold]
+    unsafe fn new(stack_size: usize) -> Result<Self, Error> {
+        let page_size = libc::sysconf(libc::_SC_PAGESIZE) as usize;
+        assert!(stack_size % page_size == 0);
+        let mem = libc::mmap(
+            std::ptr::null_mut(),
+            stack_size,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
+            -1,
+            0,
+        );
+
+        if mem == libc::MAP_FAILED || libc::mprotect(mem, page_size, libc::PROT_NONE) < 0 {
+            Err(Error::last_os_error())
+        } else {
+            assert!(mem as usize % page_size == 0);
+            let stack_id = STACK_ID.fetch_add(1, Ordering::Relaxed) as u64;
+            assert!(stack_id < MAX_STACK_ID);
+            let stack_bottom = mem.byte_add(stack_size);
+
+            // register in global variable for backtrace
+            *STACKS.get_mut()[stack_id as usize] = StackBacktrace {
+                stack_bottom,
+                stack_top: null_mut(),
+                stack_id,
+            };
+
+            Ok(Self {
+                stack_bottom,
+                stack_id,
+                stack_size,
+            })
+        }
+    }
+
+    #[inline(always)]
+    pub(super) unsafe fn record_stack(&self, stack_top: *mut c_void) {
+        let idx = self.stack_id & STACK_ID_MASK;
+        let stack = &mut STACKS.get_mut()[idx as usize];
+        stack.stack_top = stack_top;
+        stack.stack_id = self.stack_id;
+    }
+
+    #[inline(always)]
+    pub(super) unsafe fn remove_stack_record(&self) {
+        let idx = self.stack_id & STACK_ID_MASK;
+        STACKS.get_mut()[idx as usize].stack_top = null_mut();
+    }
+}
+
+unsafe impl Send for Stack {}
+unsafe impl Sync for Stack {}
+
+// Stack should never be freed
+//
+// impl Drop for Stack {
+//     #[cold]
+//     fn drop(&mut self) {
+//         assert_eq!(
+//             unsafe { libc::munmap(self.stack_bottom.byte_sub(self.stack_size), self.stack_size) },
+//             0
+//         );
+//     }
+// }
+
+struct StackPool {
+    capacity: usize,
+    stacks: Vec<Stack>,
+    global: bool,
+}
+
+impl StackPool {
+    const fn new(capacity: usize, global: bool) -> Self {
+        Self {
+            capacity,
+            stacks: Vec::new(),
+            global,
+        }
+    }
+
+    fn fetch_stack(&mut self) -> Option<Stack> {
+        self.stacks.pop()
+    }
+
+    fn return_stack(&mut self, mut stack: Stack, id_increment: u64) -> Result<(), Stack> {
+        if self.stacks.len() >= self.capacity {
+            Err(stack)
+        } else {
+            stack.stack_id += id_increment;
+            self.stacks.push(stack);
+            Ok(())
+        }
+    }
+
+    fn return_multi(&mut self, stacks: &mut Vec<Stack>) {
```

Comment:
conventionally we use "batch" rather than "multi", e.g. return_stacks_in_batch, or even return_stacks for simplicity